### DISCUSSION TITLE
Incubator database session adapter ported to 4.0?

### DISCUSSION THREAD

**Discussion by jsooter** [phalcon/cphalcon#15683] [Q&A]
Has anyone ported the database session adapter to version 4.0?

https://github.com/phalcon/incubator/blob/master/Library/Phalcon/Session/Adapter/Database.php

**Comment by wurst-hans** [comment]
What does not work using the existing adapter of Incubator? Any errors?

BTW: I've written a custom implementation based on the Incubator example, but I'm using real models instead of using native SQL for reading from and writing to database.

**Comment by jsooter** [comment]
Here is the error I get 
`Fatal error: Declaration of Phalcon\Session\Adapter\SessionDatabase::destroy($session_id = NULL) must be compatible with Phalcon\Session\Adapter\AbstractAdapter::destroy($id): bool in /var/www/project/app/library/Phalcon/SessionDatabase.php on line 169`

Here is the destroy method I have
```PHP
    public function destroy($session_id = null)
    {
        if (!$this->isStarted()) {
            return true;
        }
        if (is_null($session_id)) {
            $session_id = $this->getId();
        }
        $this->_started = false;
        $options = $this->getOptions();
        $result = $this->connection->execute(
            sprintf(
                'DELETE FROM %s WHERE %s = ?',
                $this->connection->escapeIdentifier($options['table']),
                $this->connection->escapeIdentifier($options['column_session_id'])
            ),
            [$session_id]
        );
        return $result && session_destroy();
    }
```

Source
```
    public function destroy(var id) -> bool
    {
        if !empty(id) && this->adapter->has(id) {
            return this->adapter->delete(id);
        }

        return true;
    }
```

**Reply by wurst-hans** [reply]
As you can see [here](https://github.com/phalcon/cphalcon/blob/4.1.2-release/phalcon/Session/Adapter/AbstractAdapter.zep#L34) the class methods use type declarations for return value. So just try to modify `destroy()` to `public function destroy($session_id = null): bool`, i.e. add `bool` as type hint for return value.

**Comment by jsooter** [comment]
Thanks @wurst-hans Shows how long it's been since I've written any PHP. Return type declarations are new to me 😅

**Comment by jsooter** [comment]
I don't know if I can complete this without an example. Anyone know of any examples of this anywhere?

Previously it called start() but AbstractAdapter does not have a start method. :/ 

Maybe I'm better of going back to version 3.4

**Comment by jsooter** [comment]
Yep, I'm done... going back to 3.4 or maybe another framework :(

**Reply by wurst-hans** [reply]
Sad to hear that. Switching from 3.4 to 4.0 comes with big changes in code. So session handling has been refactored completly by Phalcon team. I'm working for some years now with Phalcon and I've created my own framework based on Phalcon which I'm using then for my projects. So there is an additional framework between Phalcon and the app, because I had to extend or fix some Phalcon behavior to fit my needs. Nevertheless, I'm happy with this really small framework, even if there is much work to do and much space for improvements.

**Reply by aircodepl** [reply]
If your software is working fine and you don't need to upgrade - you shouldn't do it. If you need to do it - give yourself a break and start again in couple of days. One by one you should fix all problems, and as you go you will probably gain confidence. 

It will be a lot of refactoring, trying to understand your own code and a lot of testing, because you basically need to test everything in your app...

There is a guide which shows most of changes: https://docs.phalcon.io/4.0/en/upgrade - it's helping. But worst are classes that just don't exist anymore and you need to find replacement by looking at tutorials/code examples/phalcon code or actually rewriting your own code. Been there done that. Biggest app that I'm involved in that needs to update I decided to postpone for 2022, as I will need like 2 weeks and 100% focus on that. More important tasks are waiting...

**Comment by Jeckerson** [comment]
Yes, Session Incubator is now separate package

https://github.com/phalcon/incubator-session

**Comment by Jeckerson** [comment]
Now all incubator package for (v4) is hosted in separate repository, here is the list:
https://github.com/phalcon/incubator/tree/4.0.x#components-summary
